dev-client/src/model/sync/sync.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2024 Technology Matters
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see https://www.gnu.org/licenses/.
+ */
+
+export type ChangeRecords<T> = Record<string, ChangeRecord<T>>;
+
+export type ChangeTimestamp = number;
+
+export type ChangeRevisionId = number;
+
+export type ChangeRecord<T> = {
+  /**
+   * Unique ID for the current state since the last sync, monotonically increasing for each change.
+   * Can be used to determine whether sync responses reflect stale data.
+   * Cleared for a successful sync.
+   * Default to zero for uninitialized records.
+   */
+  revisionId?: ChangeRevisionId;
+  lastModifiedAt?: ChangeTimestamp;
+
+  lastSyncedRevisionId?: ChangeRevisionId;
+  lastSyncedData?: T;
+  lastSyncedAt?: ChangeTimestamp;
+};
+
+export const INITIAL_REVISION_ID = 0;
+
+export const getRevisionId = <T>(
+  record?: ChangeRecord<T>,
+): ChangeRevisionId => {
+  return record?.revisionId ?? INITIAL_REVISION_ID;
+};
+
+export const nextRevisionId = (
+  revisionId: ChangeRevisionId,
+): ChangeRevisionId => {
+  return revisionId + 1;
+};
+
+export const getChanges = <T>(
+  records: ChangeRecords<T>,
+  ids: string[],
+): ChangeRecords<T> => {
+  return Object.fromEntries(ids.map(id => [id, getChange(records, id)]));
+};
+
+export const getChange = <T>(
+  records: ChangeRecords<T>,
+  id: string,
+): ChangeRecord<T> => {
+  return records[id] ?? {};
+};
+
+export const markAllChanged = <T>(
+  records: ChangeRecords<T>,
```

Comment:
question: thoughts on using `Draft<ChangeRecords<T>>`? implicitly that is what is happening when this method gets called via the soil ID slice, so it seems nice to be explicit about it (since it would probably be a bug if we were using this method on something that is _not_ a draft). i'm not sure whether that would make the test code more awkward though, and don't feel too strongly about it

Reply:
I understand the principle of being more explicit about when we're operating on intermediate Redux state vs. immutable data. I'm hesitant to engineer a potentially-complex approach to that right now (as you note it could have consequences for testing) but in a subsequent PR I'd be open to it, once we fully understand how this code is used in a well-tested push implementation.